File: src/main/java/mel/wadzapi/interview/service/impl/HamStatService.java
package mel.wadzapi.interview.service.impl;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.IntStream;

import mel.wadzapi.interview.controller.dto.EventHappensDto;
import mel.wadzapi.interview.controller.dto.HamStatResponseDto;
import mel.wadzapi.interview.service.StatTimerService;
import org.springframework.stereotype.Service;

@Service
public class HamStatService implements StatTimerService {

    private Map<Instant, Map<Integer, Set<Integer>>> userStat = new ConcurrentHashMap<>();

    @Override
    public Instant today() {
        return Instant.now().truncatedTo(ChronoUnit.DAYS);
    }

    public void processEvent(EventHappensDto eventHappensDto) {
        var userMapCounts = userStat.getOrDefault(today(), new HashMap<>());
        userMapCounts.putIfAbsent(eventHappensDto.hamsterId(), new HashSet<>());
        userMapCounts.get(eventHappensDto.hamsterId()).add(eventHappensDto.wheelId());
    }

    public HamStatResponseDto calcHamStat(int... hamsterIds) {
        var lastDayMap = userStat.get(today().minus(1, ChronoUnit.DAYS));
        if (Objects.isNull(lastDayMap)) return new HamStatResponseDto(new int[0]);
        var cnt = IntStream.of(hamsterIds)
                .map(hamId -> lastDayMap.getOrDefault(hamId, Collections.emptySet()).size())
                .toArray();
        return new HamStatResponseDto(cnt);
    }

}

File: src/main/kotlin/mel/wadzapi/interview/service/impl/HamStatService.kt
package mel.wadzapi.interview.service.impl

import mel.wadzapi.interview.controller.dto.EventHappensDto
import mel.wadzapi.interview.controller.dto.HamStatResponseDto
import mel.wadzapi.interview.service.StatTimerService
import org.springframework.stereotype.Service
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*
import java.util.Collections.emptySet
import java.util.concurrent.ConcurrentHashMap
import java.util.stream.IntStream

@Service
class HamStatService : StatTimerService {
    private val userStat: Map<Instant, MutableMap<Int, MutableSet<Int>>> = ConcurrentHashMap()
    override fun today(): Instant {
        return Instant.now().truncatedTo(ChronoUnit.DAYS)
    }

    override fun processEvent(eventHappensDto: EventHappensDto) {
        val userMapCounts = userStat[today()] ?: HashMap()
        userMapCounts.putIfAbsent(eventHappensDto.hamsterId, HashSet())
        userMapCounts[eventHappensDto.hamsterId]!!.add(eventHappensDto.wheelId)
    }

    override fun calcHamStat(vararg hamsterIds: Int): HamStatResponseDto {
        val lastDayMap: Map<Int, MutableSet<Int>>? =
            userStat[today().minus(1, ChronoUnit.DAYS)]
        if (Objects.isNull(lastDayMap)) return HamStatResponseDto(IntArray(0))
        val cnt = IntStream.of(*hamsterIds)
            .map { hamId: Int -> (lastDayMap!![hamId] ?: emptySet()).size }
            .toArray()
        return HamStatResponseDto(cnt)
    }
}